### ISSUE TITLE
ADT invariant is not injected for return values

### ISSUE THREAD

**Issue by jad-hamza** [epfl-lara/stainless#559]
Issue initially found by @karnaj.
Not sure if it's the same as #309?

The assertion in `test` fails, because it does not "see" the ADT invariant from the return value of `makeWrapper`.

```scala
object ADTRequire {
  case class Wrapper(n: BigInt) {
    require(n >= 0)
  }

  def makeWrapper(x: BigInt): Wrapper = {
    if (x >= 0) Wrapper(0)
    else makeWrapper(x-1)
  }

  def test(x: BigInt) = {
    val Wrapper(n) = makeWrapper(x)
    assert(n >= 0)
  }
}
```

**Comment by samarion** [comment]
This is not an "issue" in the sense that injecting the invariant would be unsound.

Consider for example the program
```scala
def id[T](x: T): T = x
case class A(b: Boolean) {
  require(b && !id[A](this).b)
}
```
The `A` type is empty but you can prove that `inv(A(true))` holds since the solver will assume `false` when unfolding `id(A(true))` (if you inject the invariant).

However, this does raise the question of whether our approach to ADT invariants is good/useful. I describe an alternative approach in my thesis which would allow assuming the invariant on return types but it's also not very intuitive... I don't know what the ideal/best solution is.

**Comment by jad-hamza** [comment]
Could we inject these in places that are not dependencies of the ADT? (considering the ADT invariant of `A` to be a dependency of `A`). That way we could inject the `Wrapper` invariant in `test`.

**Comment by samarion** [comment]
The issue is that Inox is doing the injection, so you would need to pass an extra argument to Inox which tells it the "context" of the current expression you're solving for in order to compute the dependencies.

That's such an ugly hack that I haven't brought myself to add it to Inox...

**Comment by samarion** [comment]
Also you would have the same issue as the technique I describe in the thesis: this would lead to obscure solving because the invariant would be assumed or not based on the dependency graph.

**Comment by jad-hamza** [comment]
Yes I agree such a hack would be very strange! We could inject it directly in Stainless instead, and document it like that: you get the invariant unless you're defining the invariant itself (and functions used in the invariant).

**Comment by samarion** [comment]
If we replace invariants by refinements, then we can easily inject them in Stainless.
Otherwise, it's a bit more complicated since we may have to traverse a datastructure to inject the invariant (imagine you're returning a `List[A]` where `A` has an invariant).

I think that replacing by refinements and only getting the invariant outside its dependencies would be the cleanest approach, but users might still get hung up on the dependency restriction.

**Comment by jad-hamza** [comment]
Yes it can be quite confusing for the user and for us! When we're defining the invariant of the ADT, we're usually working on the ADT without invariant. At the frontend level, we should perhaps distinguish explicitly between these two types (the ADT without invariant, and the one with the invariant, at the top-level, and for the children recursively)? Not sure what kind of syntax we should use for that though.
